File: LeetCode/49-GroupAnagrams.js
/*

https://leetcode.com/problems/group-anagrams/description/

Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
Example 2:

Input: strs = [""]
Output: [[""]]
Example 3:

Input: strs = ["a"]
Output: [["a"]]

Constraints:

1 <= strs.length <= 104
0 <= strs[i].length <= 100
strs[i] consists of lower-case English letters.

*/

const groupAnagrams = (strings) => {
    const anagrams = {};

    strings.forEach((string, index) => {
        const sortedString = string.split('').sort().join('');

        if (anagrams[sortedString]) {
            anagrams[sortedString].push(string);
        } else {
            anagrams[sortedString] = [string];
        }
    })

    return Object.values(anagrams);
};

// var groupAnagrams = (strings) => {
//     const hash = []

//     for (let i = 0; i < strings.length; i++) {
//         const word = strings[i]
//         const sortedWord = strings[i].split("").sort().join("")

//         if (!hash[sortedWord]) {
//             hash[sortedWord] = [word]
//         } else {
//             hash[sortedWord].push(word)
//         }
//     }

//     return hash.values()
// }

File: leetcode_python/arrays_and_hashing/49_group_anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        anagrams = {}

        for word in strs:
            sorted_word = ''.join(sorted(word))
            if sorted_word not in anagrams:
                anagrams[sorted_word] = []
            anagrams[sorted_word].append(word)

        return list(anagrams.values())


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = {}

        for word in strs:
            count = [0] * 26
            for c in word:
                count[ord(c) - ord('a')] += 1

            key = tuple(count)

            if key not in anagrams:
                anagrams[key] = [word]
            else:
                anagrams[tuple(count)].append(word)

        return list(anagrams.values())
